### PR TITLE
Added script to perform collaborative filtering on 100k Movielens dataset

### PR DESCRIPTION
**Pull request by siddBanPsu**
Uses simple dot product using user and movie embedding to predict ratings of movies and used for movie recommendations.

### CODE REVIEW

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    def call(self, inputs):
+        u = self.user_embedding(inputs[:, 0])
+        ub = self.user_bias(inputs[:, 0])
+        m = self.movie_embedding(inputs[:, 1])
+        mb = self.movie_bias(inputs[:, 1])
+        x = Dot(axes=1)([u, m])
+        x = Add()([x, ub, mb])
+        x = Activation("sigmoid")(x)
+        x = Lambda(lambda x: x * (self.max_rating - self.min_rating) + self.min_rating)(
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:141]: You don't need to use a Lambda layer, you can just in-line the code:
```
x = x * (self.max_rating - self.min_rating) + self.min_rating
```

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+Date created: 2020/05/24
+Last modified: 2020/05/24
+Description: Recommending movies using user and movie embeddings trained using Movielens small dataset.
+"""
+"""
+## Introduction
+
+This example looks at
+[Collaborative filtering using Movielens dataset](https://www.kaggle.com/c/movielens-100k)
+to recommend movies to users.
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:13]: Please add a quick explanation of the model, and maybe links to external references about collaborative filtering.

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    Reshape,
+    Dot,
+    Add,
+    Embedding,
+    Activation,
+    Lambda,
+    Dense,
+)
+from tensorflow.keras.optimizers import Adam
+from tensorflow.keras.regularizers import l2
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:36]: For these 2 imports, just `from tensorflow import keras` and then use e.g. `keras.regularizers.l2`

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    "Number of users: {}, Number of Movies: {}, Min rating: {}, Max rating: {}".format(
+        n_users, n_movies, min_rating, max_rating
+    )
+)
+
+"""
+## Prepare training and validation data
+"""
+
+X = df[["user", "movie"]].values
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:76]: Use lowercase variable names, e.g. `x_train`

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 1 above)
Comment (fchollet) [reply examples/structured_data/collaborative_filtering_movielens.py:141]: Can you explain the purpose of this line?

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+
+inputs = layers.Input(shape=(2,))
+recommendation_layer = RecommenderNet(
+    n_users, n_movies, EMBEDDING_SIZE, min_rating, max_rating
+)
+outputs = recommendation_layer(inputs)
+model = keras.Model(inputs=inputs, outputs=outputs)
+opt = Adam(lr=0.001)
+model.compile(loss="mean_squared_error", optimizer=opt)
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:154]: Why use MSE instead of binary crossentropy (assuming that you normalize the movie ratings between 0 and 1)?

Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:154]: The scores are between 0.5 and 5 (minimum and maximum ratings, respectively). I use a sigmoid layer to get a number between 0 and 1, and then scale it back using the minimum and maximum ratings.

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            embeddings_regularizer=l2(1e-6),
+        )
+        self.movie_bias = Embedding(n_movies, 1)
+
+    def call(self, inputs):
+        u = self.user_embedding(inputs[:, 0])
+        ub = self.user_bias(inputs[:, 0])
+        m = self.movie_embedding(inputs[:, 1])
+        mb = self.movie_bias(inputs[:, 1])
+        x = Dot(axes=1)([u, m])
```
[the remark this answers is not in the document]
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:138]: This operation takes a long time on my mac os while the layers.Dot is very fast. It might be because of some CPU level compiler compatibility. I will use `tf.tensordot` if it still is the recommended way.

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 1 above)
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:141]: The scores are between 0.5 and 5 (minimum and maximum ratings, respectively). I use a sigmoid layer to get a number between 0 and 1, and then scale it back using the minimum and maximum ratings.

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 5 above)
Comment (fchollet) [reply examples/structured_data/collaborative_filtering_movielens.py:154]: Try it out -- regression done with NNs often works better with crossentropy.

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 1 above)
Comment (fchollet) [reply examples/structured_data/collaborative_filtering_movielens.py:141]: Better to scale the target ratings than the network's outputs (it's a lot less work and it makes the loss range more typical)

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 5 above)
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py]: Thanks @fchollet for the recommendation. I tried it and looks good. Applied scaling to the dataset. Updated code has been committed.

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+"""
+
+import tensorflow as tf
+import pandas as pd
+import numpy as np
+from tensorflow import keras
+from pathlib import Path
+from tensorflow.keras import layers
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import LabelEncoder
```
[the remark this answers is not in the document]
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py]: Thanks. Did it.

**Comment by siddBanPsu** [comment]
> Thank you for the PR! This will be a useful example :)
> 
> The code will need some work before we can approve it. Here is a first round of comments to address.

Thank you very much for providing valuable suggestions. I have incorporated most of them.

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 2 above)
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:14]: Done!

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+to recommend movies to users.
+"""
+
+import tensorflow as tf
+import pandas as pd
+import numpy as np
+from tensorflow import keras
+from pathlib import Path
+from tensorflow.keras import layers
+from sklearn.model_selection import train_test_split
```
[the remark this answers is not in the document]
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py]: Yes, I have removed sklearn altogether.

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+)
+from tensorflow.keras.optimizers import Adam
+from tensorflow.keras.regularizers import l2
+import matplotlib.pyplot as plt
+
+"""
+## First, load the data and apply some pre-processing
+"""
+
+# Download the actual data from http://files.grouplens.org/datasets/movielens/ml-latest-small.zip"
```
[the remark this answers is not in the document]
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:46]: Sure. I do an extract and remember the name of the directory that is being used for the task.

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+    def call(self, x):
+        x = self.emb(x)
+        x = Reshape((self.n_factors,))(x)
+        return x
+
+
+class RecommenderNet(keras.Model):
+    def __init__(self, n_users, n_movies, embedding_size, min_rating, max_rating):
+        super(RecommenderNet, self).__init__(name="recnet")
```
[the remark this answers is not in the document]
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py]: Thanks for the tip. I have done as you mentioned.

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+The model computes a score using dot product of user and movie embeddings.
+Thereafter, it uses a sigmoid activation on the sum over all the scores (including biases).
+Finally, the score is converted into the rating ranges based on this dataset.
+"""
+EMBEDDING_SIZE = 50
+
+
+class RecommenderNet(keras.Model):
+    def __init__(
+        self, num_users, num_movies, embedding_size, min_rating, max_rating, **kwargs
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:121]: The model no longer needs min_rating and max_rating, you can remove them

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+    ):
+        super(RecommenderNet, self).__init__(**kwargs)
+        self.num_users = num_users
+        self.num_movies = num_movies
+        self.embedding_size = embedding_size
+        self.min_rating = min_rating
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:127]: Remove

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+        self.max_rating = max_rating
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:128]: Remove

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+            embeddings_regularizer=keras.regularizers.l2(1e-6),
+        )
+        self.movie_bias = layers.Embedding(num_movies, 1)
+
+    def call(self, inputs):
+        user_vector = self.user_embedding(inputs[:, 0])
+        user_bias = self.user_bias(inputs[:, 0])
+        movie_vector = self.movie_embedding(inputs[:, 1])
+        movie_bias = self.movie_bias(inputs[:, 1])
+        dot_user_movie = layers.Dot(axes=1)([user_vector, movie_vector])
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:149]: Why not use `tf.tensordot` here?

If you want to use `Dot`, you should instantiate the layer in `__init__` and call it here (otherwise you are instantiating a new instance every time you call the model)

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+        # Add all the components (including bias)
+        x = dot_user_movie + user_bias + movie_bias
+        # The sigmoid activation forces the rating to between 0 and 1
+        x = tf.nn.sigmoid(x)
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:153]: You can put the `return` on this line

File: examples/structured_data/collaborative_filtering_movielens.py (hunk 14 above)
Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:149]: Actually tf.tensordot was taking a very long time because my conda tensorflow installation had some issues. I fixed it and will use it. Will submit the updated code soon. Thanks very much!

**Comment by fchollet** [comment]
Are you observing any performance difference between `tensordot` and `Dot`?

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+## Create the model
+
+We embed both users and movies in to 50-dimensional vectors.
+
+The model computes a match score between user and movie embeddings via a dot product,
+and adds a per-movie and per-user bias. The match score is scaled to the `[0, 1]`
+interval via a sigmoid (since our ratings are normalized to this range).
+"""
+# Setting a random seed for reproducibility
+tf.random.set_seed(1234)
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:119]: This is not meaningful; for reproducibility you to set both the global seed and per-op seeds. Since you don't do the latter, this does nothing. You can remove these 2 lines

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        dot_user_movie = tf.tensordot(user_vector, movie_vector, 2)
+        # Add all the components (including bias)
+        x = dot_user_movie + user_bias + movie_bias
+        # The sigmoid activation forces the rating to between 0 and 1
+        return tf.nn.sigmoid(x)
+
+
+inputs = layers.Input(shape=(2,))
+model = RecommenderNet(num_users, num_movies, EMBEDDING_SIZE)
+outputs = model(inputs)
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:159]: You don't need to create an input. You can just do:

```
model = RecommenderNet(num_users, num_movies, EMBEDDING_SIZE)
model.compile(...)
model.fit(...)
```

File: examples/structured_data/collaborative_filtering_movielens.py [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+movies_not_watched = [[movie2movie_encoded.get(x)] for x in movies_not_watched]
+user_encoder = user2user_encoded.get(user_id)
+user_movie_array = np.hstack(
+    ([[user_id]] * len(movies_not_watched), movies_not_watched)
+)
+ratings = model.predict(user_movie_array).flatten()
+top_ratings_indices = ratings.argsort()[-10:][::-1]
+recommended_movie_ids = [movie_encoded2movie.get(x) for x in top_ratings_indices]
+
+print("Here are the top 10 movie recommendations for user: {}".format(user_id))
```
Comment (fchollet) [inline examples/structured_data/collaborative_filtering_movielens.py:210]: Please also print the titles of the movies were rated highly by the user, so the readers can see the relationship between liked movies and the new recommendations

Comment (siddBanPsu) [reply examples/structured_data/collaborative_filtering_movielens.py:210]: Adding title + movie genre too for better understanding. Had a bug in the code when looking up movie ids that is resolved

**Comment by siddBanPsu** [comment]
> Are you observing any performance difference between `tensordot` and `Dot`?

Actually, now they are similar. However, earlier, when I was doing `Model(inputs=inputs, ...`, the training thereafter was slower than just using `model = RecommenderNet ...` and then calling `compile()` and `fit()`

**Comment by siddBanPsu** [comment]
> Looks great, thank you for the update! You can add the autogenerated files.

Thanks again @fchollet, for all the help and suggestions. I have added the remaining files.
